### ISSUE TITLE
Flickering on Live Plotting with Rich

### ISSUE THREAD

**Issue by ThSGM** [piccolomo/plotext#74]
I tried to modify the example code that was raised in [this issue](https://github.com/piccolomo/plotext/issues/26) that managed to get plotext working with Rich. This is a fairly minimal script that tries to plot a moving sine curve. 

The problem is that it shows a lot of flickering. 

I am very new to coding with Python so I'm sure there are things that I'm doing that are not right. Possible issues might be around the `def make_plot` section, or the section when I initialise the plots, or the final Live section. 

```
from rich.layout import Layout
from rich.live import Live
from rich.ansi import AnsiDecoder
from rich.console import RenderGroup
from rich.jupyter import JupyterMixin
from rich.panel import Panel

from datetime import datetime
from rich.table import Table

from time import sleep
import plotext as plt

import numpy as np

def make_plot(width, height):
    # plt.clf()
    plt.clear_data()
    plt.scatter(yy, marker="x")
    plt.plotsize(width, height)
    plt.title(str(count))
    plt.colorless()
    return plt.build()

class plotextMixin(JupyterMixin):
    def __init__(self):
        self.decoder = AnsiDecoder()
        self.count = count
        
    def __rich_console__(self, console, options):
        self.width = options.max_width or console.width
        self.height = options.height or console.height
        canvas = make_plot(self.width, self.height)
        self.rich_canvas = RenderGroup(*self.decoder.decode(canvas))
        yield self.rich_canvas

def make_layout():
    layout = Layout(name="root")
    layout.split(
        Layout(name="header", size=3),
        Layout(name="main", ratio=1),
    )
    layout["main"].split_row(
        Layout(name="plotext", size=100),
        Layout(name="main_right", size=30),
    )
    return layout

# Clear the plot
plt.clf()

# Initialise he count
count = 0

# Initialise the plots
xx = np.arange(0, 6*np.pi, 0.1)
yy = np.sin(xx)
uu = np.stack((xx,yy), axis=1)

layout = make_layout()
plotext_layout = layout["plotext"]
mix = plotextMixin()
mix = Panel(mix, style="white on black")
plotext_layout.update(mix)

header_layout = layout["header"]

with Live(layout, refresh_per_second=0.1) as live:
    while True:
        xx = xx + 0.1*count
        yy = np.sin(xx)
        uu = np.stack((xx,yy),axis=1)
        count = count + 1

        plotext_layout.update(mix)

        plt.sleep(0.01)
        plt.show()
```

**Comment by piccolomo** [comment]
Hi @ThSGM,

thanks a lot for dropping your feedback! :-)

I have modified your code and it seem to work fine on my machine. I changed colouring, marker and layout but you could change my code to suite your preferences.  Here it is:
```python
from rich.layout import Layout
from rich.live import Live
from rich.ansi import AnsiDecoder
from rich.console import RenderGroup
from rich.jupyter import JupyterMixin
from rich.panel import Panel
from rich.text import Text

from time import sleep
import plotext as plt

def make_plot(width, height, phase = 0, title = ""):
    plt.clf()
    l, frames = 1000, 30
    x = range(1, l + 1)
    y = plt.sin(1, periods = 2, length = l, phase = 2 * phase  / frames)
    plt.scatter(x, y, marker = "fhd")
    plt.plotsize(width, height)
    plt.xaxis(0, "upper")
    plt.yaxis(0, "right")
    plt.title(title)
    color = 253
    plt.canvas_color(color)
    plt.axes_color(color)
    #plt.cls()
    return plt.build()

class plotextMixin(JupyterMixin):
    def __init__(self, phase = 0, title = ""):
        self.decoder = AnsiDecoder()
        self.phase = phase
        self.title = title
        
    def __rich_console__(self, console, options):
        self.width = options.max_width or console.width
        self.height = options.height or console.height
        canvas = make_plot(self.width, self.height, self.phase, self.title)
        self.rich_canvas = RenderGroup(*self.decoder.decode(canvas))
        yield self.rich_canvas

def make_layout():
    layout = Layout(name="root")
    layout.split(
        Layout(name="header", size=1),
        Layout(name="main", ratio=1),
    )
    layout["main"].split_column(
        Layout(name="static", ratio = 1),
        Layout(name="dynamic"),
    )
    return layout

layout = make_layout()

header = layout['header']
title = plt.colorize("Pl✺ text ", "cyan bold") + "integration with " + plt.colorize("rich_", "dim")
header.update(Text(title, justify = "center"))

static = layout["static"]
phase = 0
mixin_static = Panel(plotextMixin(title = "Static Plot"))
static.update(mixin_static)

dynamic = layout["dynamic"]

with Live(layout, refresh_per_second=0.0001) as live:
    while True:
        phase += 1
        mixin_dynamic = Panel(plotextMixin(phase, "Dynamic Plot")) 
        dynamic.update(mixin_dynamic)
        #sleep(0.001)
        live.refresh()
```
which outputs a static and dynamic plot: here is a snapshot.

![image](https://user-images.githubusercontent.com/56689657/153878530-986ecf3f-b736-426e-8c19-97c21f7a71c3.png)

What do you think? Does it work for you as well? Feel free to drop your version if you find a better one :-)

All the best,
Savino

**Comment by ThSGM** [comment]
Hi!

I'm running a few tests, but the flickering is different when using different terminals on my macOS (notably Alacritty and iTerm). I'll run it on my Raspberry Pi computer to see if it persists there as well. 

In the mean, I'm wondering if you could explain a few elements of your code: 

- As I understand it, the refresh_per_second = 0.0001 is extremely slow. Basically refreshes every 10k seconds. What is the reason why you would set the Rich refresh to be so slow? Is this equivalent to [setting auto_refresh=False](https://rich.readthedocs.io/en/stable/live.html#auto-refresh)? 
- So how often would live.refresh() be called? As fast as possible? 
- How often is makeplot() run? 
- So everytime makeplot is run, the plot is cleared via `plt.clf()` -- is there an advantage to `plt_clear_data()` instead?

**Comment by piccolomo** [comment]
Hi @ThSGM ,

I am not a `rich` expert, but this is what I understood. 

- The `refresh_per_second` doesn't seem to make a difference: I tried values between 10**12 and 10**-12 with no difference, and I guess it is because there is a while loop soon after, which is what really matter. 
- I have no idea what is `auto_refresh`
- What really decides the refresh rate it seem to be the plotting time, which you could find with the function `plt.time()`. In my case, I get tens of milliseconds, so I would suggests adding `plt.sleep(t)` inside the final while loop and tweek this value to remove the flickering. Here is how:
```
with Live(layout, refresh_per_second = 0.0001) as live:
    while True:
        mixin_dynamic.phase += 1
        dynamic.update(mixin_dynamic_panel)
        plt.sleep(0.01) # tweek this time to remove flickering
        live.refresh()
```
- So ye, I guess you are right, `plt.cld()` would be sufficient. This is because only the data changes from one plot to the next and not the colouring, title, and any other setting. `plt.clf()` clears all settings (it is therefore safer), `plt.cld()` only the data plotted (it is therefore faster). 
- There is no need for `plt.show()` here  because the script uses `plt.build()` which creates the plot and passes it, as a string, to `rich`, which does the final printing. `plt.show()` would add an unnecessary further printing outside the `rich` environment.
- Btw if you want to make the plot faster, use` plt.clc()` or `plt.cls()` to remove colouring and `marker = "x"` (or any other character, except `hd` and `fhd`).

Did it work? 

Also is there any specific reason you would use `rich`? You could obtain a similar result with `plotext` alone, but with shorter code. Here is how:
```
import plotext as plt

l, frames = 1000, 30
x = range(1, l + 1)
def make_data(phase = 0):
    return plt.sin(1, periods = 2, length = l, phase = 2 * phase  / frames)

plt.clf()
plt.subplots(2, 1)

plt.subplot(1, 1)

y = make_data()
plt.scatter(x, y, marker = "fhd")
plt.title("static plot")

plt.subplot(2, 1)

for f in range(100 * frames):
    plt.cld()
    plt.clt()
    y = make_data(f)
    plt.scatter(x, y, marker = "fhd")
    plt.title("dynamic plot")
    #plt.sleep(0.01) # tweek this time to remove fleekering
    plt.show()
```
which seem a little more responsive. 

What do you think?

All the best

**Comment by piccolomo** [comment]
Hello @ThSGM, any update?

**Comment by piccolomo** [comment]
Great I will wait then for a more complete answer (but take your time). 

My latest reply could help you btw!

**Comment by ThSGM** [comment]
> Hello @ThSGM, any update?

(Re-sending on the appropriate account)

> Hello @ThSGM, any update?

Hello! Thank you for all your help. I'll do a quick reply for now as I've been busy with work. 

Basically, your alternative code did not seem to help. I tested it on both macOS iTerm2 and Alacritty and it showed significant flickering. The delay is due to the fact that I wanted to test on a Raspberry Pi/Debian system to see if it was due to the terminal. 

I have uploaded a video to show here: https://streamable.com/tl3ktr

But then, even if it does work on Linux/Debian, I think the flickering that persists on iTerm2/Alacritty is still a problem.
